standard_libs/libraries/units/source.rb

```diff
@@ -1,147 +1,156 @@
-    #
-    # @attr_reader [String] name  the bad unit name
-    class BadUnitNameError < StandardError
-        attr_reader :name
-        
-        def initialize(msg: "Unknown unit name", name:)
-            @name = name
-            super(msg)
-        end
+  EMPTY = -1
+
+  # Special Characters
+  MU = '&mu;'
+  DEGREES = '&deg;'
+
+  # Prefixes
+  PICO = 'p'
+  NANO = 'n'
+  MICRO = MU
+  MILLI = 'm'
+  KILO = 'k'
+  MEGA = 'M'
+  GIGA = 'G'
+
+  # Volume
+  LITERS = 'l'
+  MICROLITERS = MICRO + LITERS
+  MILLILITERS = MILLI + LITERS
+
+  # Weight
+  GRAMS = 'g'
+  NANOGRAMS = NANO + GRAMS
+  MICROGRAMS = MICRO + GRAMS
+
+  # Concentration
+  MOLAR = 'M'
+  PICOMOLAR = PICO + MOLAR
+  NANOMOLAR = NANO + MOLAR
+  MICROMOLAR = MICRO + MOLAR
+  MILLIMOLAR = MILLI + MOLAR
+  PERCENT = '%'
+
+  # Temperature
+  CELSIUS = 'C'
+  DEGREES_C = DEGREES + CELSIUS
+
+  # Time
+  MINUTES = 'min'
+  SECONDS = 'sec'
+  HOURS = 'hr'
+
+  # Force
+  TIMES_G = 'x g'
+
+  # R/DNA Length
+  BASEPAIRS = 'bp'
+  KILOBASEPAIRS = KILO + BASEPAIRS
+  MEGABASEPAIRS = MEGA + BASEPAIRS
+  GIGABASEPAIRS = GIGA + BASEPAIRS
+
+  # Voltage
+  VOLTS = 'V'
+  MILLIVOLTS = MILLI + VOLTS
+
+  def self.qty_display(qty)
+    sep = case qty[:units]
+          when PERCENT
+            ''
+          else
+            ' '
+          end
+    "#{qty[:qty]}#{sep}#{qty[:units]}"
+  end
+
+  def qty_display(qty)
+    Units.qty_display(qty)
+  end
+
+  def add_qty_display(options)
+    new_items = {}
+
+    options.each do |key, value|
+      key =~ /^(.+_)+([a-z]+)$/
+
+      case Regexp.last_match(2)
+      when 'microliters'
+        units = MICROLITERS
+      when 'milliliters'
+        units = MILLILITERS
+      when 'minutes'
+        units = MINUTES
+      else
+        next
+      end
+
+      qty = value.to_f
+
+      new_items["#{Regexp.last_match(1)}qty".to_sym] = { qty: qty, units: units }
     end
-    
-    # Return a key for the measure hash defined on the given object type.
-    #
-    # The measure hash must be defined in the data proerty of the object type as JSON.
-    # For instance
-    #
-    #   { "measure": { "type": "concentration", "unit": "micromolar" } }
-    #
-    # The key is constructed as the type name, an underscore, and the unit name.
-    #
-    # @param object_type [ObjectType] the object type
-    # @returns the key for the measure of the the object type if there is one
-    # @raises MissingObjectTypeMeasure if the object type has no measure data_object
-    def self.get_measure_key(object_type:)
-        data_object = object_type.data_object
-        raise MissingObjectTypeMeasureError.new(name: object_type.name) if !data_object.key?(:measure)
-        
-        measure = object_type.data_object[:measure]
-        type_name = measure[:type]
-        unit_name = measure[:unit]
-        "#{type_name}_#{self.get_unit(unit_name: unit_name)}"
+
+    options.update(new_items)
+  end
+
+  # Return the unit constant for the the unit name if there is one.
```

Comment:
@malloc3 can you check and see if this overwrites changes you made?

Reply:
This does not overwrite any changes that I have made however it does not include some of the changes I have made.   Once this is merged I can push the few things I have.